File: hs/Mod_isomorphism.hs
module Mod_isomorphism ( isomorphism ) where
import Prelude hiding (and, or) 
import Debug.Trace
import Splib
import Types
import Mod_treeToString
import Data.Char
import Mod_subtree

isomorphism :: Tree -> Tree -> (Map Node Node)
isomorphism tree1 tree2 
 | isIsomorph tree1 tree2 = joinMaps (setImage emptyMap (root tree1) (root tree2)) (isoHelper tree1 tree2 childs1 childs2 emptyMap)
 | otherwise = emptyMap
    where 
     childs1 = (children tree1 (root tree1))
     childs2 = (children tree2 (root tree2))
        
isoHelper :: Tree -> Tree -> [Node] -> [Node] -> (Map Node Node) -> (Map Node Node)
isoHelper tree1 tree2 toMatch1 toMatch2 map 
 | empty toMatch1 = map
 | isIsomorph sub1 sub2 = isoHelper tree1 tree2 (removeFirst toMatch1) (removeFirst toMatch2) (joinMaps map (isomorphism sub1 sub2))
 | otherwise = isoHelper tree1 tree2 toMatch1 shiftMatch2 map
           where
            n1 = first(toMatch1)
            n2 = first(toMatch2)
            sub1 = subtree tree1 n1
            sub2 = subtree tree2 n2
            shiftMatch2 = append (removeFirst toMatch2) (first toMatch2)

isIsomorph :: Tree -> Tree -> Bool
isIsomorph tree1 tree2 
  | isLeaf tree1 (root tree1) && isLeaf tree2 (root tree2) = True
  | length childs1 == length childs2 = isIsomorphChilds tree1 tree2 childs1 childs2 0
  | otherwise = False
    where 
     childs1 = (children tree1 (root tree1))
     childs2 = (children tree2 (root tree2))

isIsomorphChilds :: Tree -> Tree -> [Node] -> [Node] -> Int -> Bool
isIsomorphChilds tree1 tree2 toMatch1 toMatch2 i
 | empty toMatch1 = True
 | i > length toMatch2 = False
 | isIsomorph sub1 sub2 = isIsomorphChilds tree1 tree2 (removeFirst toMatch1) (removeFirst toMatch2) 0
 | otherwise = isIsomorphChilds tree1 tree2 toMatch1 shiftMatch2 (i+1)
           where
            n1 = first(toMatch1)
            n2 = first(toMatch2)
            sub1 = subtree tree1 n1
            sub2 = subtree tree2 n2
            shiftMatch2 = append (removeFirst toMatch2) (first toMatch2)

File: cpp/Mod_isomorphism.cpp
#pragma once

bool isIsomorph(Tree t1, Tree t2) {
    if(isLeaf(t1, root(t1)) && isLeaf(t2, root(t2))) {
        return true;
    }
    
    if (length(children(t1, root(t1))) == length(children(t2, root(t2)))) {
        List<Node> toMatch1 = children(t1, root(t1));
        List<Node> toMatch2 = children(t2, root(t2));
        
        int cnt = 0;
        while(!empty(toMatch1)){
            Node n1 = first(toMatch1);
            Node n2 = first(toMatch2);
            if(cnt > length(toMatch2)){
                return false;
            }
            
            if(isIsomorph(subtree(t1, n1), subtree(t2, n2))){
                toMatch1 = removeFirst(toMatch1);
                toMatch2 = removeFirst(toMatch2);
                cnt = 0;
            }else{
                toMatch2 = append(removeFirst(toMatch2), first(toMatch2));
                cnt++;
            }
        }
        return true;
    }else{
        return false;
    }
}

Map<Node, Node> mapIsomorph(Tree t1, Tree t2, Map<Node, Node> map) {
    map = setImage(map, root(t1), root(t2));
    
    List<Node> toMatch1 = children(t1, root(t1));
    List<Node> toMatch2 = children(t2, root(t2));
        
    while(!empty(toMatch1)){
        Node n1 = first(toMatch1);
        Node n2 = first(toMatch2);
        
        if(isIsomorph(subtree(t1, n1), subtree(t2, n2))){
           toMatch1 = removeFirst(toMatch1);
           toMatch2 = removeFirst(toMatch2);
           map = mapIsomorph(subtree(t1, n1), subtree(t2, n2), map);
        }else{
            toMatch2 = append(removeFirst(toMatch2), first(toMatch2));
        }
    }
    return map;
}

Map<Node, Node> isomorphism(Tree t1, Tree t2) {
    Map<Node, Node> map = emptyMap<Node, Node>();
    
    if(isIsomorph(t1, t2)) {
        return mapIsomorph(t1,t2, map);
    }
    
    return emptyMap<Node, Node>();
        
}
